### ISSUE TITLE
twisted.vfs.adapters.ftp._FileToConsumerAdapter doesn't fully implement IConsumer

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2735]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2735|
|Type|defect|
|Created|2007-07-16 10:17:47Z|

_FileToConsumerAdapter pretends to implement IConsumer, but doesn't define registerProducer and unregisterProducer.

I don't see how this could ever work. Maybe it didn't?

<details><summary>Searchable metadata</summary>

```
trac-id__2735 2735
type__defect defect
reporter__therve therve
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__vfs vfs
keywords__ 
time__1184581067000000 1184581067000000
changetime__1195122150000000 1195122150000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="KyleJ's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|KyleJ commented|
|-|-|

This worked for me:

```
class _FileToConsumerAdapter:
    implements(IConsumer)

    producer = None
    producerIsStreaming = None

    def __init__(self, original):
        self.original = original
        self.offset = 0

    def write(self, bytes):
        self.original.writeChunk(self.offset, bytes)
        self.offset += len(bytes)

    def registerProducer(self, producer, streaming):
        self.producer = producer
        self.producerIsStreaming = streaming

    def unregisterProducer(self):
        if self.producer is not None:
            del self.producer
            del self.producerIsStreaming
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set status to `closed`|
|-|-|

(In [21220]) Merge vfs-ftp-1264+2735

Author: therve
Reviewers: cablehead, exarkun
Fixes [#7212](https://github.com/twisted/twisted/issues/7212)
Fixes [#2735](https://github.com/twisted/twisted/issues/2735)

Add tests for FTP interfaces IFTPShell, IReadFile and IWriteFile. This led to
some cleanups in ftp.FTPShell, and removed several bugs in vfs ftp adapter and
inmem backend.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set status to `closed`|
|-|-|

(In [21793]) Merge vfs-ftp-1264+2735-2

Author: therve
Reviewers: jml, exarkun, cablehead
Fixes [#7212](https://github.com/twisted/twisted/issues/7212)
Fixes [#2735](https://github.com/twisted/twisted/issues/2735)

Refactor tests for the vfs ftp adapter, using the new IFTPShellTestsMixin,
and fixe behavior both in the ftp adapter and in the inmem backend.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
